refactor: psbt uxui

fbwoolf opened the pull request:
> Try out this version of the Hiro Wallet - download [extension builds](https://github.com/hirosystems/wallet/actions/runs/5658817419).<!-- Sticky Header Marker -->

This PR refactors the PSBT code to address UI changes (new designs), and aims to cleanup the previous implementation. In addition to showing the inputs/outputs with tags and the raw tx, we now total transfer/receiving (with inscriptions) and show the fee. This work also adds tags and warnings based on sighash types.

I have also addressed catching and routing the user to an error page if the signer lib errors on decoding or signing.

![Screen Shot 2023-07-17 at 1 49 50 PM](https://github.com/hirosystems/wallet/assets/6493321/47407a43-de4e-46d2-b6f9-93c8723949c7)

![Screen Shot 2023-07-17 at 1 50 08 PM](https://github.com/hirosystems/wallet/assets/6493321/7440378a-7ea7-45a6-b061-ea8c420ab1cf)

![Screen Shot 2023-07-17 at 1 50 21 PM](https://github.com/hirosystems/wallet/assets/6493321/8332b992-b951-4e6c-b65d-736079e006a2)

![Screen Shot 2023-07-17 at 1 50 38 PM](https://github.com/hirosystems/wallet/assets/6493321/f8e0242f-72f3-4c4a-a3f5-a7c849e4d3f4)

![Screen Shot 2023-07-17 at 1 50 52 PM](https://github.com/hirosystems/wallet/assets/6493321/0460f064-cdae-48cd-b9fa-206cf7190999)

![Screen Shot 2023-07-17 at 1 51 10 PM](https://github.com/hirosystems/wallet/assets/6493321/0fbcc761-8df4-4aa0-b9e2-10b3b492434c)

![Screen Shot 2023-07-17 at 1 51 21 PM](https://github.com/hirosystems/wallet/assets/6493321/45e8944c-3f4c-4c7b-83fc-61dffe217c6e)

![Screen Shot 2023-07-17 at 3 20 03 PM](https://github.com/hirosystems/wallet/assets/6493321/1d055d96-a963-436f-9285-b77448ff16f5)

alter-eggo commented on src/app/features/psbt-signer/components/psbt-request-details/components/psbt-inputs-and-outputs/psbt-inputs-and-outputs.tsx:
```diff
@@ -0,0 +1,49 @@
+}: PsbtInputsAndOutputsProps) {
+  const [showDetails, setShowDetails] = useState(false);
+
+  if (!inputs.length && !outputs.length) return null;
```
don't we expect both inputs and outputs here? shouldn't it be 
```
if (!inputs.length || !outputs.length)
```
?

fbwoolf replied:
Yep, good point, it prob should be on `or`. Thanks.

alter-eggo commented on src/app/features/psbt-signer/hooks/use-parsed-psbt.tsx:
```diff
@@ -0,0 +1,103 @@
+  const matchedSatsWithInscriptions = useMemo(() => {
+    return matchedSatsWithOutputs.matchedSats.filter(match =>
+      match.inputs.find(utxo => {
+        if (isDefined(utxo.inscription)) return utxo.inscription;
```
cannot it be just 
```
return utxo.inscription
```

fbwoolf replied:
Yep, haha, thanks. Not sure how I ended up with that one. 🙃

alter-eggo commented on src/shared/crypto/bitcoin/bitcoin.utils.ts:
```diff
@@ -61,36 +59,48 @@ export function decodeBitcoinTx(tx: string) {
+      type: outputScript.type,
+      pubkey: outputScript.pubkey,
+    });
+  } else if (outputScript.type === 'ms' || outputScript.type === 'tr_ms') {
```
always feel like just using `if` is better than `else if` in such code. but up to you here

fbwoolf replied:
Ok, yep, I'll refactor it. 👍

alter-eggo commented on src/app/features/psbt-signer/hooks/match-sats-with-outputs.ts:
```diff
@@ -0,0 +1,84 @@
+  return Object.values(utxo);
+}
+
+export function matchSatsWithOuputs(
```
maybe it would be better to add some test for this function? it would help to understand what this function does

fbwoolf replied:
I do have a test included to test the matching ...in same name spec file.

fbwoolf replied:
I can look at how to add more tests to it though to breakdown what is happening.

alter-eggo reviewed (commented):
great work @fbwoolf !

fbwoolf commented on src/app/features/psbt-signer/components/psbt-request-details/components/psbt-inputs-and-outputs/components/psbt-input-output-item.tsx:
```diff
@@ -0,0 +1,104 @@
+  txId?: string;
+  txIdHoverLabel?: string;
+}
+export function PsbtInputOutputItem({
```
True.

fbwoolf commented on src/app/features/psbt-signer/components/psbt-request-details/components/psbt-inputs-and-outputs/components/psbt-output-list/components/psbt-output-list.layout.tsx:
```diff
@@ -0,0 +1,11 @@
+
+export function PsbtOutputListLayout({ children }: HasChildren) {
+  return (
+    <Box>
+      <Box borderLeft="1px solid #DCDDE2">{children}</Box>
+    </Box>
+  );
```
Not quite sure, let me check.

fbwoolf replied:
If I remove...

![Screen Shot 2023-07-19 at 11 32 02 AM](https://github.com/hirosystems/wallet/assets/6493321/b005ebdb-7dea-4a48-8205-8af70aec48fc)

fbwoolf commented on src/app/features/psbt-signer/components/psbt-request-details/components/psbt-inputs-outputs-totals/components/psbt-address-total-item.tsx:
```diff
@@ -1,41 +1,43 @@
-          {value}
+          {title ? title : 'Bitcoin'}
         </Text>
+        <Box _hover={{ cursor: 'pointer' }} as="button" onClick={valueAction} type="button">
```
I saw that too and meant to remove. I copied from some other code.

fbwoolf commented on src/app/features/psbt-signer/components/psbt-request-details/psbt-request-details.tsx:
```diff
@@ -0,0 +1,90 @@
+}
+
+function getPsbtTxOutputs(psbtTx: btc.Transaction) {
+  const outputsLength = psbtTx.outputsLength;
+  const outputs: btc.TransactionOutput[] = [];
+  if (outputsLength === 0) return outputs;
+  for (let i = 0; i < outputsLength; i++) {
+    outputs.push(psbtTx.getOutput(i));
+  }
+  return outputs;
```
Good times.

pete-watters replied:
Is that why we use a `for` instead of a `map` here? 🤔

fbwoolf commented on src/app/query/bitcoin/ordinals/ordinals-aware-utxo.query.ts:
```diff
@@ -36,7 +36,7 @@ export async function getNumberOfInscriptionOnUtxo(id: string, index: number) {
 async function fetchOrdinalsAwareUtxo(
   txid: string,
-  index: number
+  index?: number
```
The type from btc-signer lib has the index as possible undefined so it has to be handled somewhere. Doesn't work to handle it in the `useQueries`.

kyranjamie replied:
Can we handle elsewhere rather than changing the schema of this fn? Maybe a function: `getIndexWithDefaultZero`?

fbwoolf replied:
Why would we force the index to be zero? Wouldn't that be problematic to assume?

fbwoolf replied:
I think you would want the query to fail if the index is undefined? Not force a default?

kyranjamie replied:
I see this as a leaky abstraction. Because of btc-signer specific implementation problem, other parts of the code are updated to reflect said issue.

But a utxo has both a txid and an index. There's no reason to ever call it without one. So a better solution is to prevent it being called in the first place, rather than update the function to allow illegal cases.

fbwoolf replied:
Yeah, I agree, I'll figure out a way to catch it earlier.

fbwoolf commented on src/app/features/psbt-signer/hooks/match-sats-with-outputs.ts:
```diff
@@ -0,0 +1,84 @@
+    matchedSatsWithOutput.push(matched);
+  }
+
+  const matchedSats: MatchedSatsWithOutput[] = matchedSatsWithOutput.map(matched => {
+    return {
+      inputs: consolidateSatsWithSameAddress(matched.inputs),
+      output: matched.output,
+    };
+  });
+
+  return { fee: remainingSats, matchedSats };
```
Sats are the value on the input (utxo) ...makes sense to me. Ultimately, we need to know the inputs/outputs that match up bc the input (utxo) funds the output and the tx output is what we query to check for the inscription. I can change the name to use `inputs`? Sats are the input value.

Not sure I understand your suggestion?

fbwoolf replied:
I think maybe I just have the logic wrong in how inputs and outputs work? I was thinking a utxo is used across multiple outputs until it is spent, but that isn't correct? It can only be used for one output. Probably why doesn't make sense to you? I think I get your suggestion now.

kyranjamie replied:
> Sats are the value on the input

Maybe we have a different understanding.

I'd describe as: An input is a collection of sats. Afaik, the only reason we ever need to "match", is to know if an inscription (on a sat) goes to a given output. 

If so, then isn't it better to think of it as where does this sat (of this input) go, rather than where this input goes. 
This is a simpler model as a sat-to-output is a one to one relationship, whereas an input to output is a one to many relationship.

fbwoolf replied:
I think what you describe is what I'm doing, I am using the input value (sats) to do the matching. However, I don't think the input/output is a one-to-many relationship? That is what I'm questioning now about my implementation. I was just doing more reading on it, and I think one input (utxo) is matched to one output. Are you confident input/output is one-to-many bc chatGPT tells me differently... the remainder of each utxo is change.

<img width="843" alt="Screen Shot 2023-07-19 at 12 06 05 PM" src="https://github.com/hirosystems/wallet/assets/6493321/41e633db-91d5-4245-91db-84242e2b507f">

fbwoolf replied:
And, yes, we are currently using just for inscriptions and the fee (but we can determine the fee with a simple subtraction of outputs from inputs if needed).

fbwoolf replied:
Redoing bc I over-complicated the matching, brb!

fbwoolf replied:
Actually, I don't think I did. I think chapGPT is incorrect? It depends really on the sighash type?

fbwoolf replied:
> I'd describe as: An input is a collection of sats. Afaik, the only reason we ever need to "match", is to know if an inscription (on a sat) goes to a given output.
> 
We don't make a query per sat for the inscription. I am matching up the sat value of the input with the outputs to determine if that utxo is being spent for that output. Then, I'm checking if the utxo itself has an inscription, not per sat.
>
> If so, then isn't it better to think of it as where does this sat (of this input) go, rather than where this input goes. This is a simpler model as a sat-to-output is a one to one relationship, whereas an input to output is a one to many relationship.
> 
I think this is what I'm doing tbh? Except that, if the sighash type isn't ALL, there would be a problem.
> 
> Am I correct in thinking we only "match" to know where an inscription goes? or are there other use cases?
>
Yes, and currently for the fee.

fbwoolf replied:
> 
> const output = satMatcher.findOutput({ txid, offset });
> ```
> 
I don't understand this, how are you finding a tx output on a PSBT with a `txid` and an `offset`? The output on a PSBT is a `script` (which contains the receiving address) and an `amount`. 🤔

fbwoolf replied:
Also, I followed a similar pattern I saw others doing, where I parse the PSBT inputs and put some extra data on it to use in the UI, type `PsbtInput`, this already has the inscription data, if there is one. So, when they are matched to outputs, it is easy to know if there is an inscription on the transferring `input` (utxo) or receiving `output`.

fbwoolf replied:
I did some renaming today, and added some more tests, to help clarify what is happening here. Although it isn't checking inscriptions per sat, I do feel like it is working at the utxo level to detect the inscription and which output is receiving it. I think this makes the assumption the tx is being constructed correctly though.

fbwoolf replied:
Yep, I get everything you are saying and agree. I don't think what you are suggesting is far off what I have done. The totals are being done separately. The only prob is the issue with the example you show. I think there is just a gap in my understanding when you say `finds corresponding output that has it` ...I have no idea what you mean there if you don't do matching?

kyranjamie replied:
Another example to better highlight what we discussed on call

<img width="910" alt="image" src="https://github.com/hirosystems/wallet/assets/1618764/bc0222d7-6235-40b7-9ebc-456f289c7f76">

pete-watters commented on src/app/common/psbt/use-psbt-request-params.ts:
```diff
@@ -0,0 +1,48 @@
+  const requestToken = searchParams.get('request');
+  return useMemo(
+    () => ({
+      origin,
```
Does the tabId of `1` or `0` mean something? Maybe we could put them in a well named `enum`? 

```
enum PSBT_TABS: { 
 I_AM: 0,
A_TAB: 1
}
```

kyranjamie replied:
@pete-watters tabId is a auto-generated value from the browser we don't know upfront

pete-watters replied:
OK, so it's browser tab, thanks!

pete-watters commented on src/app/features/psbt-signer/components/psbt-request-details/components/psbt-inputs-outputs-totals/components/psbt-address-totals.tsx:
```diff
@@ -0,0 +1,67 @@
+  const accountHasOutputInscriptions = accountOutputsReceivingInscriptions?.length;
+
+  return (
+    <>
```
Maybe we can use `&&` here and avoid the `?  <> : null` 🤔

fbwoolf replied:
I do think in our code we tend to pref either an early return or the ternary `null` over `&&` when rendering components. @kyranjamie am I wrong here?

kyranjamie replied:
Prefer early return where possible. Don't really have a pref on ternary or &&. Would it be worth standardising?

pete-watters commented on src/app/features/psbt-signer/components/psbt-request-details/components/psbt-inputs-outputs-totals/psbt-inputs-outputs-totals.tsx:
```diff
@@ -0,0 +1,76 @@
+  const isReceiving = isNativeSegwitTotalLessThanZero || isTaprootTotalLessThanZero;
+  const showDivider = isTransferring && isReceiving;
+
+  if (addressNativeSegwitTotal.amount.isEqualTo(0) && addressTaprootTotal.amount.isEqualTo(0))
```
How about a const for total zero like above? 

```
 const totalZero = addressNativeSegwitTotal.amount.isEqualTo(0) && addressTaprootTotal.amount.isEqualTo(0));
 
 if(totalZero) return null;
```

pete-watters commented on src/app/features/psbt-signer/components/psbt-request-details/components/psbt-request-fee.tsx:
```diff
@@ -0,0 +1,32 @@
+  return (
+    <PsbtRequestDetailsSectionLayout>
+      <SpaceBetween>
+        <Text fontSize={2} fontWeight="500">
```
We could extract it to it's own component to avoid repeating.

In general throughout the APP we have a lot of repeated text and styles. When implementing the new design system we should consider:
- implementing tokens to reuse for styles
- extracting text to `i18n` / config files

pete-watters commented on src/app/features/psbt-signer/components/psbt-request-details/components/psbt-request-raw.tsx:
```diff
@@ -0,0 +1,50 @@
+        title="Raw transaction"
+      />
+      {showDetails ? (
+        <Box
```
How about making a component for this? It seems useful and could be re-used

pete-watters commented on src/app/features/psbt-signer/components/psbt-request-sighash-warning-label.tsx:
```diff
@@ -0,0 +1,15 @@
+  return (
+    <WarningLabel title="Be careful with this transaction" width="100%">
+      This is a potentially dangerous. In agreeing to proceed, you allow another part to change the
+      content of the transaction later, therefore the details you see now are not guaranteed.{' '}
```
Can we use padding instead of `{' '}` ?

pete-watters commented on src/app/features/psbt-signer/hooks/match-utxos-with-outputs.ts:
```diff
@@ -0,0 +1,72 @@
+// import { isDefined } from '@shared/utils';
```
Can ditch this commented code?

pete-watters commented on src/app/routes/app-routes.tsx:
```diff
@@ -86,6 +87,88 @@ function useAppRoutes() {
+      <Route
+        path={RouteUrls.PsbtRequest}
+        element={
+          <AccountGate>
```
Maybe we can extend `AccountGate` to accept a prop for a suspense wrapper? 

Then we won't need to repeat the `<Suspense fallback={<LoadingSpinner height="600px" />}>` so often

fbwoolf replied:
Good suggestion, I'll try to take on tackling that here.

fbwoolf replied:
@pete-watters can you make an issue for this? This refactor is quite large already and this code was just moved, so I'm hoping to avoid taking this on here. You can use the label on the issue `code health`.

pete-watters replied:
Sure, no problem. I agree his is already big enough 👍 #4051

pete-watters commented on src/shared/crypto/bitcoin/bitcoin.utils.ts:
```diff
@@ -61,36 +59,49 @@ export function decodeBitcoinTx(tx: string) {
-  else if (outScript.type === 'sh') return formatKey(script, [bitcoinNetwork?.scriptHash]);
-  logger.error(`Unknown address type=${outScript.type}`);
-  return '';
+// Outscript types (btc-signer lib):
```
Maybe we can add these in a types file?

fbwoolf replied:
Yep, I can do that as temp solution. I really need to put up an issue in the lib and ask him to export the types.

kyranjamie replied:
make a PR on the repo to export this

fbwoolf replied:
This was really left as a ref, do we really need the types here?

pete-watters reviewed (commented):
Great work on this @fbwoolf 👏 

I added some comments as I went through but nothing serious, just questions / suggestions

markmhx commented:
I'm seeing that the UI for approving the "Setup transaction" on Gamma is showing the fees value both at the top under the "You'll transfer" area and also at the bottom under "Transaction fee".

My sense here is that it'd be cleanest to show fees just under "Transaction fee" and not under "You'll transfer" as well (i.e. the amount should be deducted from the "You'll transfer" total, and if it results in zero, then that area indicates no transfers).

@mica000 is that your sense as well?

<img width="1362" alt="Screenshot 2023-07-21 at 17 23 13" src="https://github.com/hirosystems/wallet/assets/28991/ae2ef3c9-4380-49b4-b778-3100c2568b57">

markmhx commented:
I'm also seeing the "Requested by" line is indented to the right but there's no favicon shown in that area. Should we align to the left sans indentation? And can we show the domain in addition to the app name for security assurances? (e.g. `Gamma (gamma.io)`

<img width="455" alt=" " src="https://github.com/hirosystems/wallet/assets/28991/f881f38c-2a65-450e-b0bd-6cb6bdd30243">

markmhx commented:
> Please review the recipient address, amount, and associated fees. Authorize only transactions you fully understand.

@mica000 I'm wondering whether we need this copy or whether the need to review and understand the details displayed below is implicit?

In any case, it seems more accurate and simpler to say "Please review this transaction's recipients, amounts and fees before confirming."

Also, having "Approve transaction" up top then "Transaction" a bit further down seems a bit redundant. Could we remove the second "Transaction" line and put the "Certain" pill to the right of "Approve transaction" to simplify?

<img width="468" alt="Screenshot 2023-07-21 at 17 33 17" src="https://github.com/hirosystems/wallet/assets/28991/00e9fe15-c1a6-4cc5-9901-13801497264b">

markmhx commented:
It appears we need to wire up the tooltips for "Approve" and "You" under inputs/outputs still:

<img width="237" alt="Screenshot 2023-07-21 at 17 33 41" src="https://github.com/hirosystems/wallet/assets/28991/db74612a-b95b-4713-b3f1-dca4c0791ae4">

markmhx commented:
For the Gamma "purchase transaction" transaction, I see it's marked as "Uncertain":

<img width="442" alt="Screenshot 2023-07-21 at 17 34 59" src="https://github.com/hirosystems/wallet/assets/28991/c3ab3e40-74ca-463a-9348-f1ecf3335979">

> This is a potentially dangerous. In agreeing to proceed, you allow another part to change the content of the transaction later, therefore the details you see now are not guaranteed

Update to: `The details you see here are not guaranteed. Be sure to fully trust your counterparty, who can later modify this transaction to send or receive other assets from your account, and possibly even drain it.`

> More about signatures types on Bitcoin

Update to: `Learn more` and link to user guide about sigtypes -> @314159265359879 do we have this already?

The "Uncertain" pill also needs a tooltip still.

markmhx commented:
Re: the favicon, it appears we may be querying the API with the site name instead of domain?

<img width="1116" alt="Screenshot 2023-07-21 at 17 53 49" src="https://github.com/hirosystems/wallet/assets/28991/36b0ee60-ba0f-4074-bcad-edb2a60926b0">

markmhx commented:
On Magic Eden, I'm finding that the transfer/receive data is showing up as expected but the loading spinners are also appearing perpetually regardless:

<img width="1365" alt="Screenshot 2023-07-21 at 17 53 14" src="https://github.com/hirosystems/wallet/assets/28991/03fcb4cb-744e-4619-a9d8-e1c8f3ac603a">

fbwoolf commented:
> My sense here is that it'd be cleanest to show fees just under "Transaction fee" and not under "You'll transfer" as well (i.e. the amount should be deducted from the "You'll transfer" total, and if it results in zero, then that area indicates no transfers).
> 
Not sure I agree here. 🤔 Typically wallets are showing the `You'll transfer` including the fee, and no fee separated. Technically, it is part of what you will be transferring.

fbwoolf commented:
> I'm also seeing the "Requested by" line is indented to the right but there's no favicon shown in that area. Should we align to the left sans indentation? And can we show the domain in addition to the app name for security assurances? (e.g. `Gamma (gamma.io)`
> 
This is weird, I always see the favicon. In the designs, I think it is indented further like shown here.

![Screen Shot 2023-07-21 at 12 36 49 PM](https://github.com/hirosystems/wallet/assets/6493321/0d460378-38ed-41b0-9a70-57e2af6e466b)

fbwoolf commented:
> Re: the favicon, it appears we may be querying the API with the site name instead of domain?
> 
This broke with the rpc work, I fixed it and changed the warning text...

![Screen Shot 2023-07-21 at 2 25 34 PM](https://github.com/hirosystems/wallet/assets/6493321/80aeb3f6-65a8-4703-ae1f-0d00691e84d2)

fbwoolf commented:
> It appears we need to wire up the tooltips for "Approve" and "You" under inputs/outputs still:
> 
Tooltips have been added...

![Screen Shot 2023-07-21 at 2 36 33 PM](https://github.com/hirosystems/wallet/assets/6493321/c6def285-2cad-409f-88e7-a883a6f174e2)

![Screen Shot 2023-07-21 at 2 44 09 PM](https://github.com/hirosystems/wallet/assets/6493321/c7c3f53e-537f-4ed6-9704-fcb2c6e903a8)

![Screen Shot 2023-07-21 at 2 45 23 PM](https://github.com/hirosystems/wallet/assets/6493321/bab4708c-8b71-47de-9ef0-9d85a4e49e54)

fbwoolf commented:
@kyranjamie can you review the changes made here in `find-outputs-receiving-inscriptions.ts` to see if the logic makes sense for what you were thinking. I wrote two tests, one with the scenario you posted ^ here, and I believe the logic is correct. Also, I added the default `tapInternalKey` changes here with some improvements to `signPsbt` that I hope simplify things by just using `signIndex` (check `usePsbtSigner`). Thanks!

kyranjamie commented on src/app/pages/psbt-request/use-psbt-request.tsx:
```diff
@@ -1,68 +1,83 @@
+        try {
+          return getDecodedPsbt(payload.hex);
+        } catch (e) {
+          return navigate(RouteUrls.RequestError, {
+            state: { message: e instanceof Error ? e.message : '', title: 'Failed request' },
+          });
```
I wouldn't put side effects like this in a getter. Can we do a separate check somewhere to make sure we can decode the psbt?

kyranjamie commented on src/app/features/psbt-signer/hooks/use-psbt-signer.tsx:
```diff
@@ -20,26 +30,36 @@ export function usePsbtSigner() {
-            logger.error('Error signing PSBT', e1, e2);
-          }
+          inputs.forEach((input, idx) => {
+            const userToSign = isUndefined(indexesToSign) || indexesToSign.includes(idx);
```
`userToSign` is an odd name imo

kyranjamie commented on src/app/features/psbt-signer/hooks/find-outputs-receiving-inscriptions.ts:
```diff
@@ -0,0 +1,49 @@
+// If an input has an inscription, we use the offset to determine where it matches
+// up from total input sats position to total output sats position. By doing this,
+// we can predict where the inscription will be sent.
+export function findOutputsReceivingInscriptions({
+  inscriptions,
+  psbtInputs,
+  psbtOutputs,
+}: FindOutputsReceivingInscriptionsArgs) {
+  let inputsSatsTotal = new BigNumber(0);
+
+  return psbtInputs
+    .flatMap(input => {
+      if (input.inscription) {
+        const inscription = inscriptions.find(inscription =>
+          input.inscription?.includes(inscription.id)
+        );
+
+        // Offset is zero indexed, so 1 is added here to match the sats total
+        const inscriptionTotalOffset = inputsSatsTotal.plus(Number(inscription?.offset) + 1);
+
+        let outputsSatsTotal = new BigNumber(0);
+
+        // Add up all the output sats until the inscription is included in the output total.
+        // This should also detect if an inscription is being lost to fees bc the outputs will
+        // never add up to include the inscription, therefore it won't be shown.
+        for (let output = 0; output < psbtOutputs.length; output++) {
+          outputsSatsTotal = outputsSatsTotal.plus(psbtOutputs[output].value);
+          if (inscriptionTotalOffset.isLessThanOrEqualTo(outputsSatsTotal))
+            return { address: psbtOutputs[output].address, inscription: input.inscription };
+        }
+      }
+      inputsSatsTotal = inputsSatsTotal.plus(input.value);
+      return;
+    })
+    .filter(isDefined);
+}
```
LGTM, though wonder if logic could better be divided such that it's two functions, where one just takes a single inscription, and gets the output. and if you want this resulting data structure, you have another fn that loops an array of inscriptions.

Signature:
```ts
findOutputReceivingInscription(inscription, inputs, outputs) // takes single inscription
```

fbwoolf replied:
I realized I don't understand what you are asking for here?

kyranjamie commented on src/app/pages/psbt-request/use-psbt-request.tsx:
```diff
@@ -1,68 +1,83 @@
     const payload = getPsbtPayloadFromToken(requestToken);
     const payloadTxBytes = hexToBytes(payload.hex);
     const appName = payload?.appDetails?.name;
+    const allowedSighash = undefinedIfLengthZero(
+      payload?.allowedSighash?.map(h => Number(h)) as btc.SignatureHash[]
+    );
+    const indexesToSign = undefinedIfLengthZero(
+      ensureArray(payload?.signAtIndex)?.map(h => Number(h))
+    );
```
Isn't this repeating logic that takes place in the request params hooks?

fbwoolf replied:
I think what you are referencing is the rpc request params? The legacy ones were always handled diff in all the legacy requests. I can look at moving them though.

fbwoolf replied:
Yeah, looked, I'll just move the legacy params into the hook like the rpc request does so handled the same. Just have to decode the `requestToken` there.

kyranjamie commented on src/app/components/json.tsx:
```diff
@@ -0,0 +1,30 @@
+import { isTypedArray } from '@shared/utils';
+
+function parseJsonReadable(value: any) {
+  if (typeof value === 'bigint') return value.toString();
```
could be a helper with the others, isString etc

fbwoolf replied:
You had actually already made this. 🤣

kyranjamie commented on src/app/features/psbt-signer/components/psbt-request-details/components/psbt-inputs-and-outputs/components/psbt-output-list/components/psbt-output-item.tsx:
```diff
@@ -0,0 +1,34 @@
+  output,
+}: PsbtOutputItemProps) {
+  const isOutputCurrentAddress =
+    output.address === addressNativeSegwit || output.address === addressTaproot;
```
How can we share `isCurrentAccountBitcoinAddress` logic? We have this repeated many places

kyranjamie commented on src/app/features/psbt-signer/components/psbt-request-details/components/psbt-inputs-and-outputs/components/psbt-output-list/components/psbt-output-item.tsx:
```diff
@@ -0,0 +1,34 @@
+      address={truncateMiddle(output.address)}
+      addressHoverLabel={output.address}
+      amount={formatMoney(createMoney(Number(output.value), 'BTC'))}
+      pillHoverLabel={isOutputCurrentAddress ? pillHoverLabel : undefined}
+      pillLabel={isOutputCurrentAddress ? <>You</> : undefined}
```
Why not pass in a pill component, rather than creating a configuration via props?

```tsx
<PsbtInputOutputItemLayout
  label={<Pill label={isOutputCurrentAddress ? pillHoverLabel : undefined} />}
  ...
/>
```

kyranjamie commented on src/app/features/psbt-signer/hooks/use-parsed-outputs.tsx:
```diff
@@ -0,0 +1,54 @@
+          return {
+            address: outputAddress,
+            mutable: isPsbtMutable,
+            sign: isCurrentAddress,
```
Wonder if we can choose a better name, `sign` is a verb and that doesn't map well to a boolean imo. Maybe `toSign`?

kyranjamie commented on src/app/pages/rpc-sign-psbt/rpc-sign-psbt.tsx:
```diff
@@ -78,14 +56,24 @@ function useRpcSignPsbt() {
+      origin={origin}
+      onSignPsbt={onSignPsbt}
+      onCancel={onCancel}
+      psbtRaw={decodedPsbt}
+      psbtTx={tx}
```
Do we need to pass in both? One can be derived from the other, no?

fbwoolf replied:
They are both derived in different getters from the passed psbt hex. They aren't derived from each other. So, I can just pass psbt hex itself and get what I need later if clearer?

markmhx commented:
Re: favicon, I see it's trying to query with the domain correctly now but still getting a CSP error:

![Screenshot 2023-07-24 at 11 25 57](https://github.com/hirosystems/wallet/assets/28991/581fe7f0-eb88-4093-a679-32f6a5e572cd)

markmhx commented:
Re: tooltips, it seems as though the content for the Certain/Uncertain labels are inverted e.g. here it shows content more appropriate for the uncertain label:

![Screenshot 2023-07-24 at 11 30 03](https://github.com/hirosystems/wallet/assets/28991/54659498-8ec7-4295-a123-ded75f515dea)

markmhx commented:
For the "Approve" tooltip in inputs/outputs specifically, "Your approval is needed to complete this transaction" is a bit ambiguous to me. It makes me think I'll need to take a subsequent step or something.

Might it be clearer as "Value you're approving as contribution to transaction".

kyranjamie reviewed (commented):
Great work Fara, big simplification in removing the matching.

Added some comments, curious to hear your take, but otherwise think this is ready to go. 

Architecturally speaking, there is a lot of prop passing going on, I wonder if this might benefit from a container/context parent component, passing certain values down to two or three smart child components.

Something in the vein of 

<img width="765" alt="image" src="https://github.com/hirosystems/wallet/assets/1618764/b6315dab-a1a3-478f-a244-a875daf1f9dc">

This might be one way to reduce the amount of repetition to figure out if it's a user address.

Not necessary to take on these changes, prop passing isn't a big deal here, just contemplating ways logic might more easily be shared across the feature.

kyranjamie commented on src/app/features/psbt-signer/components/psbt-request-details/components/psbt-inputs-and-outputs/components/psbt-input-list/components/psbt-input-item.tsx:
```diff
@@ -0,0 +1,26 @@
+import { truncateMiddle } from '@stacks/ui-utils';
```
Just on folder structure

<img width="314" alt="image" src="https://github.com/hirosystems/wallet/assets/1618764/41589710-1904-46ba-83e6-fae905b61c0f">

Is there a logical way to flatten this out? Odd to have `components/../components/../components`.

pete-watters replied:
I think that's actually good as then it's grouping the components for that area there. 

Sometimes I would promote something big like `psbt-inputs-and-outputs` to being a `feature` of `psbt-request-details` (like how we use `pages` )

fbwoolf replied:
I do like keeping the `components` folder as I have it. I think it makes the folder structure easier to read and seems more organized. Good point, @pete-watters, about the feature. I'll think on that one to see if will work here.

fbwoolf commented:
> Re: favicon, I see it's trying to query with the domain correctly now but still getting a CSP error:
> 
Your internet must be really slow? You pickup on some things I don't ever see in the UI. I'll have to throttle my loading. I suppose we should just have a default icon here bc I see the Gamma icon?

kyranjamie commented:
@fbwoolf the production CSP is tighter than the development one, so maybe that's the issue? Probs not internet speed

fbwoolf commented:
> Not necessary to take on these changes, prop passing isn't a big deal here, just contemplating ways logic might more easily be shared across the feature.
>
Agree here, but I looked back and realized I was already handling this example specifically on the parsed inputs and outputs to use in the UI, so I already had `toSign` available on each input/output to work with. I'll continue to look at refining this code though in my next PSBT issue.

fbwoolf commented:
> @fbwoolf the production CSP is tighter than the development one, so maybe that's the issue? Probs not internet speed

@kyranjamie can you look at the last commit to check the changes to the csp in prod to make the favicon work? Source:
https://content-security-policy.com/img-src/

kyranjamie commented:
> @kyranjamie can you look at the last commit to check the changes to the csp in prod to make the favicon work? Source:
https://content-security-policy.com/img-src/

The solution is more simple. My mistake. We point to a `http` url for the favicon. It should be `https`, this is already permitted by the CSP.

fbwoolf commented:
> > @kyranjamie can you look at the last commit to check the changes to the csp in prod to make the favicon work? Source:
> > https://content-security-policy.com/img-src/
> 
> The solution is more simple. My mistake. We point to a `http` url for the favicon. It should be `https`, this is already permitted by the CSP.

Got it, I see that now, thx.
